File: ios/StableChannels/StableChannels/Services/WebSocket/ProcessedTxStore.swift
import Foundation
import os.log

/// Store for recently-processed WebSocket txids with TTL and memory cap.
@MainActor
final class ProcessedTxStore {
    var entries: [String: Date] = [:]
    let ttl: TimeInterval
    let maxEntries: Int

    private var insertionOrder: [String] = []
    private var lastPurgeTime: Date = .distantPast
    private let purgeInterval: TimeInterval = 300 // 5 minutes
    private let logger = Logger(subsystem: "com.stablechannels", category: "dedup")

    init(ttl: TimeInterval = 900, maxEntries: Int = 500) {
        self.ttl = ttl
        self.maxEntries = maxEntries
    }

    // MARK: - Public API

    /// Returns `true` if the key was seen within the TTL window.
    func isRecentlyProcessed(_ key: String) -> Bool {
        guard let lastSeen = entries[key] else { return false }
        return Date().timeIntervalSince(lastSeen) < ttl
    }

    /// Record a processed key and enforce the memory cap.
    func recordProcessedTx(_ key: String) {
        if entries[key] == nil {
            insertionOrder.append(key)
        }
        entries[key] = Date()
        enforceCap()
        purgeExpiredIfDue()
    }

    // MARK: - Internal (testable)

    /// Enforce a hard entry cap by evicting the oldest 20% of entries.
    func enforceCap() {
        guard entries.count > maxEntries else { return }
        let evictCount = max(maxEntries / 5, 1) // remove ~20%
        var evicted = 0
        var writeIdx = 0
        for i in 0..<insertionOrder.count {
            let key = insertionOrder[i]
            if evicted < evictCount {
                if entries.removeValue(forKey: key) != nil {
                    evicted += 1
                }
            } else if entries[key] != nil {
                insertionOrder[writeIdx] = key
                writeIdx += 1
            }
        }
        insertionOrder.removeSubrange(writeIdx..<insertionOrder.count)
        logger.debug("Evicted \(evictCount) oldest entries (count was \(self.entries.count))")
    }

    /// Purge expired entries at most every 5 minutes.
    func purgeExpiredIfDue() {
        let now = Date()
        if now.timeIntervalSince(lastPurgeTime) < purgeInterval {
            return
        }
        let cutoff = now.timeIntervalSince1970 - ttl
        let before = entries.count
        entries = entries.filter { _, date in
            date.timeIntervalSince1970 > cutoff
        }
        let removed = before - entries.count
        if removed > 0 {
            insertionOrder.removeAll { self.entries[$0] == nil }
            logger.debug("Purged \(removed) expired entries")
        }
        lastPurgeTime = now
    }

    /// Current entry count (for tests and metrics).
    var count: Int { entries.count }
}

File: android/app/src/main/java/com/stablechannels/app/services/websocket/ProcessedTxStore.kt
package com.stablechannels.app.services.websocket

class ProcessedTxStore(
    private val ttlMs: Long = 900_000L,
    private val maxEntries: Int = 500,
) {
    private val entries = LinkedHashMap<String, Long>()
    private val lock = Any()

    @Volatile private var lastPurgeAtMs: Long = 0L

    private val purgeIntervalMs: Long = 300_000L

    fun isRecentlyProcessed(key: String): Boolean {
        val seenAt = synchronized(lock) { entries[key] } ?: return false
        return (System.currentTimeMillis() - seenAt) < ttlMs
    }

    fun recordProcessedTx(key: String) {
        val now = System.currentTimeMillis()
        synchronized(lock) {
            entries.remove(key)
            entries[key] = now
            enforceCap()
            purgeExpiredIfDue(now)
        }
    }

    fun count(): Int = synchronized(lock) { entries.size }

    private fun enforceCap() {
        if (entries.size <= maxEntries) {
            return
        }
        val evictCount = (maxEntries / 5).coerceAtLeast(1)
        val it = entries.keys.iterator()
        var removed = 0
        while (it.hasNext() && removed < evictCount) {
            it.next()
            it.remove()
            removed++
        }
    }

    private fun purgeExpiredIfDue(now: Long) {
        if ((now - lastPurgeAtMs) < purgeIntervalMs) {
            return
        }
        val cutoff = now - ttlMs
        entries.values.removeIf { it <= cutoff }
        lastPurgeAtMs = now
    }
}
